### ISSUE TITLE
Allow searching of forum posts

### ISSUE THREAD

**Issue by irilyth** [buttonmen-dev/buttonmen#1825]
It'd be nice if you could search for a string, and get back a list of forum posts containing it. It should probably work across boards, but it might also be nice if you could sort/restrict/something it by board.

**Comment by AdmiralJota** [comment]
So, some thoughts...

We probably don't want the search UI to be cluttering up every forum page. Would we want to integrate the search into the main forum page (the one that lists the boards), or make it a separate "page"? If the latter, we probably want to link to it from the main header, since I don't think there's a good consistent spot from within the forum pages.

Even if it's treated as a separate "page", I don't think it needs to actually be a separate .html or .js file, not least since we'll probably want to share code (such as formatting code) from the main forum pages. Which means we'll need to pass some kind of parameter to distinguish it. Unfortunately for @cgolubi1, unless we want to do a major re-write right now, that might mean a hashbang parameter.

Do we want search results to appear as threads or posts? Probably posts, but I think that including the title of the thread as part of the search result item would be a good thing.Probably some hybrid of what we currently use for the list of threads and what we use for the posts in a thread might be the best view. We'd want to color-code them, of course, but it might be good to include the name of the forum that each matching post was in as actual text, too.

For the search options, I'm thinking there are six fields it might be useful to search on: thread title, post text, board, poster, posted date and edited date. But for the UI, four/five inputs should suffice: text, board, poster and date range (technically two inputs).

All fields optional, but at least <i>one</i> of text, poster, start date or end date has to be provided. Fields are restrictive (so if you supply text <i>and</i> a poster, results have to match both). For text, I think it might make sense for it to match the post text for any post, and to also match the thread title, but <i>only</i> for the first post in the thread (so if your search happens to match the title of a popular thread, you don't get every single post in your result set). For board and poster, match on the board and poster, of course. For date range, I think it'd make sense if it checked both the posted date and the edit date and allowed it if either one matched the range. And of course, if you only supply one end of the range

As for text, do we want to break it down by words (so searching for "skill stripped skills" would bring up the post suggesting the Antidote skill, even though that literal string doesn't appear in it)? Or treat it as just a literal substring (so that "ing an Antidote d" would match "...using an Antidote die..."? And how should we handle the markup? It might be odd if you couldn't find a phrase that included an italicized word unless you actually specified the "[i]" inside your search string.

What do other people think of those ideas?

**Comment by blackshadowshade** [comment]
What about adding the search as a search box to the header when a forum is showing? Or were you thinking of having a proper advanced search facility, which will need its own page?

I think that supporting literal strings is important. I'm not convinced about literal substrings, though.

One question that I have, which is related to forums and displaying results, is whether we currently have any facility to divide up the results into different pages. I've noticed that the current forum overviews are getting rather long, because there doesn't seem to be any maximum number of results per page.

**Comment by AdmiralJota** [comment]
I was thinking of a full advanced search -- which I suppose we <i>could</i> fit into the main Forum page header, if we tried really hard: 
![bm1](https://cloud.githubusercontent.com/assets/5848173/11506807/3163b634-981f-11e5-80f0-a691bafa4ac3.png)

Or we could just do some subset of those things -- I suppose the date is probably the least important, and dropping that might make it small enough to potentially fit on the other forum pages as well.

As for pagination, we don't have that facility yet. It seemed like a low priority when the forum was first being implemented, and something that could be added later when it became necessary. I suppose that might be now.

**Comment by AdmiralJota** [comment]
(As a side note, dropping the dates and fitting everything into the left side of the table -- i.e., not using the space above the "45 threads" blocks -- would make the HTML/CSS for this <i>far</i> simpler, due to some weird table-rendering idiosyncrasies. So that's another reason to leave out the date range if we do decide to stuff an advanced search into the header.)

**Comment by CarltonN** [comment]
I also agree date is probably the least important of the search terms. Usually, I am wanting search the forums it more like ' where's the post about how to do X' or where was that post on Y. More often than not it's a tangential post fifteen or so in and I have no recollection of the topic. ;{D}

**Comment by TheOrgg** [comment]
Trying to find the thread where "Fuzzy" dice were concepted for the Cryptid set, I'd like to say this'd be a handy feature.

**Comment by TheOrgg** [comment]
Once again searching for a thread I remember where Cycozar explained their Fanatic, but cannot find it.  This tool'd be cool!
